### PR TITLE
Chapter 11. Account activation

### PR DESCRIPTION
**Pull request by devtuananh**

### CODE REVIEW

File: app/controllers/account_activations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class AccountActivationsController < ApplicationController
+
+  def edit
+    user = User.find_by email: params[:email]
+    if user && !user.activated? && user.authenticated? :activation, params[:id]
+      user.update_attributes activated: true, activated_at: Time.zone.now
+      user.activate
+      log_in user
+        flash[:success] = t(".activated")
```
Comment (mdmahmudurrahman) [inline app/controllers/account_activations_controller.rb]: 1. Indentation
2. remove ()

File: app/controllers/account_activations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        redirect_to user
+      else
+        flash[:danger] = t(".invalid")
```
Comment (mdmahmudurrahman) [inline app/controllers/account_activations_controller.rb]: remove ()

File: app/models/user.rb [hunk 3]
```diff
@@ -1,14 +1,38 @@
...
+
   validates :name, presence: true, length: {maximum: Settings.validates.name.maximum}
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   validates :email, presence: true, length: {maximum: Settings.validates.email.maximum},
     format: {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false}
   has_secure_password
-  validates :password, presence: true, length: {minimum: Settings.validates.password.minimum}, 
-    allow_nil: true
+  validates :password, presence: true, length:
+    {minimum: Settings.validates.password.minimum},allow_nil: true
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: space between parameters

File: app/models/user.rb [hunk 4]
```diff
@@ -1,14 +1,38 @@
...
+
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: space

File: app/views/layouts/mailer.html.erb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
+    <style>
+      /* Email styles need to be inline */
```
Comment (mdmahmudurrahman) [inline app/views/layouts/mailer.html.erb]: remove

File: app/views/user_mailer/account_activation.html.erb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<h1><%= t "user_mailer.nameapp" %></h1>
+
+<p>Hi <%= @user.name %>,</p>
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/account_activation.html.erb]: I18n

File: app/views/user_mailer/password_reset.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<h1>User#password_reset</h1>
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/password_reset.html.erb]: I18n

File: app/views/user_mailer/password_reset.html.erb [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+
+<p>
+  <%= @greeting %>, find me in app/views/user_mailer/password_reset.html.erb
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/password_reset.html.erb]: I18n

File: app/views/user_mailer/password_reset.text.erb [hunk 9]
```diff
@@ -0,0 +1,3 @@
+User#password_reset
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/password_reset.text.erb]: I18n

File: app/views/user_mailer/password_reset.text.erb [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+
+<%= @greeting %>, find me in app/views/user_mailer/password_reset.text.erb
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/password_reset.text.erb]: same here

File: app/views/user_mailer/password_reset.html.erb (hunk 7 above)
Comment (devtuananh) [reply app/views/user_mailer/password_reset.html.erb]: Yes

File: config/environments/production.rb [hunk 11]
```diff
@@ -1,94 +1,39 @@
...
-
-  # Use a real queuing backend for Active Job (and separate queues per environment)
-  # config.active_job.queue_adapter     = :resque
-  # config.active_job.queue_name_prefix = "sample_app_#{Rails.env}"
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.delivery_method = :smtp
+  host = "sample-app-tuan-anh.herokuapp.com"
+  config.action_mailer.default_url_options = { host: host }
+  ActionMailer::Base.smtp_settings = {
+    :address        => "smtp.sendgrid.net",
```
Comment (mdmahmudurrahman) [inline config/environments/production.rb]: Hash racket syntax is deprecated, use symbol.

File: app/views/user_mailer/password_reset.text.erb (hunk 10 above)
Comment (devtuananh) [reply app/views/user_mailer/password_reset.text.erb]: Ok, I edited it

File: config/environments/production.rb (hunk 11 above)
Comment (devtuananh) [reply config/environments/production.rb]: Yes

Comment (devtuananh) [reply config/environments/production.rb]: ActionMailer::Base.smtp_settings = {:address, :port, :authentication,          :user_name, :password, :domain, :enable_starttls_auto}
  => {"smtp.sendgrid.net", "587", :plain, ENV["SENDGRID_USERNAME"],
       ENV["SENDGRID_PASSWORD"], "heroku.com", true}

Comment (devtuananh) [reply config/environments/production.rb]: I write it correct, don't you?

File: app/models/user.rb [hunk 12]
```diff
@@ -1,14 +1,40 @@
...
+  before_create :create_activation_digest
+
+  validates :name, presence: true, length:
+    {maximum: Settings.validates.name.maximum}
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-  validates :email, presence: true, length: {maximum: Settings.validates.email.maximum},
-    format: {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false}
+  validates :email, presence: true, length:
+    {maximum: Settings.validates.email.maximum}, format:
+      {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false}
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: As you already used 2 space right alignment in the previous line, now for this line you don't need 2 more space here. Just align it with it's previous line

File: app/models/user.rb [hunk 13]
```diff
@@ -1,14 +1,40 @@
...
+    {minimum: Settings.validates.password.minimum}, allow_nil: true
+
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password? token
+  end
+
+  def remember
+    self.remember_token = User.new_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove self

File: app/models/user.rb [hunk 14]
```diff
@@ -22,16 +48,14 @@ def new_token
...
-  def remember
-    self.remember_token = User.new_token
-    update_attributes remember_digest: User.digest(remember_token)
-  end
+  private
 
-  def authenticated?(remember_token)
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def downcase_email
+    self.email = email.downcase
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove self

File: app/models/user.rb [hunk 15]
```diff
@@ -22,16 +48,14 @@ def new_token
...
   end
 
-  def forget
-    update_attributes remember_digest: User.digest(remember_token)
+  def create_activation_digest
+    self.activation_token  = User.new_token
+    self.activation_digest = User.digest activation_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove self

File: app/models/user.rb [hunk 16]
```diff
@@ -22,16 +48,14 @@ def new_token
...
-  def authenticated?(remember_token)
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def downcase_email
+    self.email = email.downcase
   end
 
-  def forget
-    update_attributes remember_digest: User.digest(remember_token)
+  def create_activation_digest
+    self.activation_token  = User.new_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove self

File: app/views/layouts/mailer.html.erb [hunk 17]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
+    <style>
+
```
Comment (mdmahmudurrahman) [inline app/views/layouts/mailer.html.erb]: remove empty line

File: config/environments/development.rb [hunk 18]
```diff
@@ -1,61 +1,33 @@
...
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
+      address: "smtp.gmail.com",
+      port: 587,
+      domain: "gmail.com",
+      authentication: "plain",
+      enable_starttls_auto: true,
+      user_name: "trananh9a@gmail.com",
+      password: "Hanoivn1"
```
Comment (mdmahmudurrahman) [inline config/environments/development.rb]: 1. 2 space indentation
2. Place the gmail and password in the environment variable, else hacker will do what ever he like =))

Comment (devtuananh) [reply config/environments/development.rb]: Yes, I know it :)) I use gem Figaro, really

File: app/models/user.rb [hunk 19]
```diff
@@ -1,14 +1,40 @@
 class User < ApplicationRecord
-  
-  attr_accessor :remember_token
-  before_save {email.downcase!}
-  validates :name, presence: true, length: {maximum: Settings.validates.name.maximum}
+  attr_accessor :remember_token, :activation_token
+  before_save   :downcase_email
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1 space

Comment (devtuananh) [reply app/models/user.rb]: ýes

File: app/models/user.rb [hunk 20]
```diff
@@ -22,16 +48,14 @@ def new_token
...
-  def remember
-    self.remember_token = User.new_token
-    update_attributes remember_digest: User.digest(remember_token)
-  end
+  private
 
-  def authenticated?(remember_token)
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def downcase_email
+    email = email.downcase
```
Comment (hungnh103) [inline app/models/user.rb]: `email.downcase!`

File: app/models/user.rb [hunk 21]
```diff
@@ -22,16 +48,14 @@ def new_token
...
   end
 
-  def forget
-    update_attributes remember_digest: User.digest(remember_token)
+  def create_activation_digest
+    activation_token  = User.new_token
```
Comment (hungnh103) [inline app/models/user.rb]: remove redundant whitespace

File: app/views/user_mailer/account_activation.text.erb [hunk 22]
```diff
@@ -0,0 +1,3 @@
+<%= t "user_mailer.hi" %> <%= @user.name %>,
+<%= t "user_mailer.account_span" %>
+<%= edit_account_activation_url(@user.activation_token, email: @user.email) %>
```
Comment (hungnh103) [inline app/views/user_mailer/account_activation.text.erb]: remove `( )`

File: config/routes.rb [hunk 23]
```diff
@@ -10,6 +10,7 @@
   get "/login", to: "sessions#new"
   post "/login", to: "sessions#create"
   delete "/logout", to: "sessions#destroy"
-  
+
   resources :users
+  resources :account_activations, only: [:edit]
```
Comment (hungnh103) [inline config/routes.rb]: `only: :edit`

**Pull request merged**